File: smartdevices-app/smartdevicesapp.model/src/main/java/de/vogler_engineering/smartdevicesapp/model/entities/DeviceId.java
/**
 * Copyright (c) Vogler Engineering GmbH. All rights reserved.
 * Licensed under the MIT License. See LICENSE.md in the project root for license information.
 */
package de.vogler_engineering.smartdevicesapp.model.entities;

import com.fasterxml.jackson.annotation.JsonIgnore;

import lombok.Data;

/**
 * Created by vh on 14.02.2018.
 */

@Data
public class DeviceId {

    private static String ID_KEY_SPACER = ".";

    private final String username;
    private final String deviceId;

    @JsonIgnore
    public String getIdKey(){
        return String.format("%s%s%s", username, ID_KEY_SPACER, deviceId);
    }

    public DeviceId(String username, String deviceId){
        this.username = username;
        this.deviceId = deviceId;
    }

    public DeviceId(String idKey){
        int dot = idKey.indexOf(ID_KEY_SPACER);
        if(dot <= 0 || dot == idKey.length()-1 || idKey.indexOf(ID_KEY_SPACER, dot+1) != -1)
            throw new IllegalArgumentException("Cannot parse DeviceIdKey");
        this.username = idKey.substring(0, dot);
        this.deviceId = idKey.substring(dot+1);
    }
}

File: smartdevices-gateway/SmartDevicesGateway.Model/Dto/Config/DeviceId.cs
//
// Copyright (c) Vogler Engineering GmbH. All rights reserved.
// Licensed under the MIT License. See LICENSE.md in the project root for license information.
//
using System;
using System.Collections.Generic;
using System.Text;

namespace SmartDevicesGateway.Model.Dto.Config
{
    public class DeviceId : IComparable<DeviceId>
    {
        public DeviceId()
        {
            //for litedb
        }

        public const string Separator = ".";

        public string User { get; set; }

        public string DeviceName { get; set; }

        public string FullId => User + Separator + DeviceName;

        public DeviceId(string user, string deviceName)
        {
            User = user;
            DeviceName = deviceName;
        }

        public DeviceId(string id)
        {
            var i = id.IndexOf(Separator, StringComparison.Ordinal);
            if (i < 0 || i == id.Length)
            {
                throw new FormatException("Cant read device ID");
            }

            User = id.Substring(0, i);
            DeviceName = id.Substring(i + 1);
        }

        public int CompareTo(DeviceId other)
        {
            return string.CompareOrdinal(FullId, other.FullId);
//
//
//            var i = string.Compare(User, other.User, StringComparison.Ordinal);
//            return i != 0 ? i : string.Compare(DeviceName, other.DeviceName, StringComparison.Ordinal);
        }

        public override bool Equals(object obj)
        {
            if (!(obj is DeviceId))
            {
                return false;
            }

            var other = (DeviceId) obj;

            return FullId.Equals(other.FullId);
        }


        public override int GetHashCode()
        {
            return Tuple.Create(User, DeviceName).GetHashCode();
        }

        public override string ToString()
        {
            return FullId;
        }

        public static DeviceId TryParse(string deviceId)
        {
            if (deviceId == null)
            {
                return null;
            }
            try
            {
                return new DeviceId(deviceId);
            }
            catch (FormatException)
            {
                return null;
            }
        }
    }
}
